File: examples/go/math-demo/MathDemo.tsop.go
package contract

import "tsop"

type MathDemo struct {
	tsop.StatefulSmartContract
	Value tsop.Bigint
}

func (c *MathDemo) DivideBy(divisor tsop.Bigint) {
	c.Value = tsop.Safediv(c.Value, divisor)
}

func (c *MathDemo) WithdrawWithFee(amount, feeBps tsop.Bigint) {
	fee := tsop.PercentOf(amount, feeBps)
	total := amount + fee
	tsop.Assert(total <= c.Value)
	c.Value = c.Value - total
}

func (c *MathDemo) ClampValue(lo, hi tsop.Bigint) {
	c.Value = tsop.Clamp(c.Value, lo, hi)
}

func (c *MathDemo) Normalize() {
	c.Value = tsop.Sign(c.Value)
}

func (c *MathDemo) Exponentiate(exp tsop.Bigint) {
	c.Value = tsop.Pow(c.Value, exp)
}

func (c *MathDemo) SquareRoot() {
	c.Value = tsop.Sqrt(c.Value)
}

func (c *MathDemo) ReduceGcd(other tsop.Bigint) {
	c.Value = tsop.Gcd(c.Value, other)
}

func (c *MathDemo) ScaleByRatio(numerator, denominator tsop.Bigint) {
	c.Value = tsop.MulDiv(c.Value, numerator, denominator)
}

func (c *MathDemo) ComputeLog2() {
	c.Value = tsop.Log2(c.Value)
}

File: examples/rust/math-demo/MathDemo.tsop.rs
use tsop::prelude::*;

#[tsop::contract]
pub struct MathDemo {
    pub value: Bigint,
}

#[tsop::methods(MathDemo)]
impl MathDemo {
    #[public]
    pub fn divide_by(&mut self, divisor: Bigint) {
        self.value = safediv(self.value, divisor);
    }

    #[public]
    pub fn withdraw_with_fee(&mut self, amount: Bigint, fee_bps: Bigint) {
        let fee = percent_of(amount, fee_bps);
        let total = amount + fee;
        assert!(total <= self.value);
        self.value = self.value - total;
    }

    #[public]
    pub fn clamp_value(&mut self, lo: Bigint, hi: Bigint) {
        self.value = clamp(self.value, lo, hi);
    }

    #[public]
    pub fn normalize(&mut self) {
        self.value = sign(self.value);
    }

    #[public]
    pub fn exponentiate(&mut self, exp: Bigint) {
        self.value = pow(self.value, exp);
    }

    #[public]
    pub fn square_root(&mut self) {
        self.value = sqrt(self.value);
    }

    #[public]
    pub fn reduce_gcd(&mut self, other: Bigint) {
        self.value = gcd(self.value, other);
    }

    #[public]
    pub fn scale_by_ratio(&mut self, numerator: Bigint, denominator: Bigint) {
        self.value = mul_div(self.value, numerator, denominator);
    }

    #[public]
    pub fn compute_log2(&mut self) {
        self.value = log2(self.value);
    }
}
